spec/helpers.lua

```diff
@@ -168,6 +178,10 @@ local function get_db_utils(strategy, tables)
     assert(dao:run_migrations())
   end
 
+  db:truncate("plugins")
+
+  assert(plugins_loader.load_plugins(conf, db))
```

Comment:
It does what is says, it's there to load the plugins :) ...along with their schemas.

In the old-DAO implementation of Plugins, the schemas were loaded on-demand during the self.check function, with the following now-removed code (note the comment):

```
    -- singletons.configuration would be nil when plugin operations are
    -- done through DAOs like in migrations or tests
    if singletons.configuration and not singletons.configuration.loaded_plugins[plugin_name] then
      return nil, "plugin '" .. plugin_name .. "' not enabled; " ..
                  "add it to the 'plugins' configuration property"
    end

    local loaded, plugin_schema = utils.load_module_if_exists("kong.plugins."
                                    .. plugin_name .. ".schema")
    if loaded then
      return plugin_schema
    else
      return nil, 'Plugin "' .. tostring(plugin_name) .. '" not found'
    end
```

Migrations no longer use DAOs, and tests now load plugins explicitly via this `load_plugins` call in `get_db_utils` (note also that `helpers.get_db_utils` now has an additional optional argument `plugins`, used when tests need to load plugins other than the default set).

If you still think this would benefit from a comment I can add it, though I can't think of anything better than `-- load plugins along with their schemas`. I think this `load_plugins` call is in line with what the rest of this function does (emulate `Kong.init`), so in that regard it doesn't strike me as too special (though I see how it could be puzzling given old-DAO assumptions). If anything, `get_db_utils` should eventually be renamed `init` :)

Reply:
> is in line with what the rest of this function does (emulate Kong.init, roughly speaking

That's... really not how I see this function. I don't really see it being renamed to `init` either. To me, it instantiates the DAO and that is all. Why are plugins schemas not loaded from within the DAO itself? `load_plugins` is to load plugins _configurations_ (rows in the `plugins` table), but the schemas is a DAO responsibility, since they are DAO extensions.

Reply:
This is the sort of side-effects I was referring to in my above comment that makes me think this is some sort of design flaw as an unaware reader

Reply:
> That's... really not how I see this function. I don't really see it being renamed to init either. To me, it instantiates the DAO and that is all. load_plugins is to load plugins configurations (rows in the plugins table), but the schemas is a DAO responsibility, since they are DAO extensions.

I'll agree on the first point, that `init` is not the name or role of `get_db_utils` (in spite of the similarities, `Kong.init` is really not a good comparison), but I'll disagree on the second point about `load_plugins`.

The only reason the `load_plugins` function in the current `master`/`next` ever reads rows in the `plugins` table is to give an error in case some plugin in the DB is not in the config file (in fact that could be split into a separate function). After that, those rows are never used, and the meat of the function's effect is to load the `handler` and `schema` modules of the statically configured plugins and produce their execution order. So what `plugin_loader.load_plugins` — the refactored version of that function — does is the same, so it really loads the plugins, as in, it makes the system aware of each plugin's existence (as an entity). So yeah, the goal of `load_plugins` is mostly to produce a side-effect on the `db` argument (to load plugins into it) and to return the ordered list.

In fact, `load_plugins` could be split into `check_plugins_in_db` (which would read rows), `load_plugin_entities_into_db` (which would load the subschemas into the db) and `sort_plugins` (which would take the list of entities and produce an order for execution. In `get_db_utils` we only really need that middle part. And come to think of it, this middle part could even be a custom method in the Plugins entity DAO.

Well, writing this comment was useful self-rubber-ducking :) I'll try to quickly split these three parts and see how it looks like, I think the end result will become nicer and less confusing.